### DISCUSSION TITLE
Is there a detailed tutorial on "websocket"?

### DISCUSSION THREAD

**Discussion by hgalytoby** [permitio/fastapi_websocket_pubsub#47] [General]
Sorry, my native language is not English, my English is poor, if communication is not good I will close issues.

I found the websocket example from https://github.com/permitio/fastapi_websocket_pubsub/issues/41

Before I found this example, it was not clear that the message should be sent after connecting.

### Subscribe Msg
 ```
{
    request:  {
        method: "subscribe",
        "arguments": {
        "topics": ["event"]
        }
    }
}
```

I understand that I can receive messages by sending `Subscribe Msg`.

I understand sending messages to topics using Python.
```py
endpoint = PubSubEndpoint(broadcaster="redis://localhost:6379")
await endpoint.publish(['event'], {'msg': 'ok'})
```

How can I send a message to a topic from Web JS?

Do you have any examples or documentation to look at?

### 1.I want to send a message to `Event topic`?
```js
let ws = new WebSocket("ws://localhost:8080/pubsub")
ws.send(JSON.stringify(????))
```

### 2.After the web receives a message once, there are no more messages.
```py
import asyncio
from fastapi_websocket_pubsub import PubSubEndpoint


async def main():
    endpoint = PubSubEndpoint(broadcaster='redis://localhost:6379')
    while True:
        await endpoint.publish(["event"], data={'msg': 'hello web'})
        await asyncio.sleep(1)


if __name__ == '__main__':
    asyncio.run(main())
```

But the web can receive the trigger message.
```py
async def events():
    await endpoint.publish(['event'], {"msg": "fastapi"})


@app.get('/trigger')
async def trigger_events():
    asyncio.create_task(events())
```

![image](https://user-images.githubusercontent.com/50397689/186202655-86874cf1-dfaa-46c2-a2c4-6ca49e43975d.png)

**Comment by orweis** [comment]
Hi, and sorry for the delayed reply- had a busy  week.
I think your english is pretty clear, I understand you want to write a Javascript client to connect with the websocket.
This is the same as this discussion: https://github.com/permitio/fastapi_websocket_pubsub/discussions/42 (Which you also linked to).

In order to have the JS client work it **must** implement the [RPC protocol(https://github.com/permitio/fastapi_websocket_rpc/blob/master/fastapi_websocket_rpc/rpc_channel.py), which is another logical layer on top of the websocket.

The bear minimum here is to implement the right [schema for an RPC message](https://github.com/permitio/fastapi_websocket_rpc/blob/8c7fd8947b3a993b81976467b437a9fb945a83eb/fastapi_websocket_rpc/schemas.py#L1) and call the pub/sub [RPC method publish](https://github.com/permitio/fastapi_websocket_pubsub/blob/49b8dfc7eac055d264ea492d559cee975d5574a6/fastapi_websocket_pubsub/rpc_event_methods.py#L47)

Which is something like this:
```JSON
{
 "request":  {
      "method": "publish",  
      "arguments": {"topics":["event"], "data":{"msg": "hello web"} }, 
      "call_id": "<RANDOM-UUID>" 
  } 
}
```

**Comment by hgalytoby** [comment]
Thank you for your reply, my English is translated by Google.

I later flipped through the source code and found publish
https://github.com/permitio/fastapi_websocket_pubsub/blob/49b8dfc7eac055d264ea492d559cee975d5574a6/fastapi_websocket_pubsub/rpc_event_methods.py#L47

I have some questions now, but I'm having a hard time expressing in English.

I have attached a gif, can you take a look?

![20220827_015022](https://user-images.githubusercontent.com/50397689/186963747-fb141422-237d-49a7-9bd8-73b6d5618cab.gif)

Why web js subscribe after...

Client.py can't receive the message of test.py?

Then I used fastapi trigger with no problem.

But when I add a web connection.

There is no way to receive fastapi trigger messages.

After I disconnected the initial web connection, I was able to receive messages.

What am I doing wrong? Or am I using the method wrong?

### server.py
```python
import asyncio
import uvicorn
from fastapi import FastAPI
from fastapi.routing import APIRouter
from fastapi_websocket_pubsub import PubSubEndpoint
from starlette.websockets import WebSocket

app = FastAPI()
router = APIRouter()
endpoint = PubSubEndpoint(broadcaster='redis://localhost:6379')


@router.websocket('/pubsub')
async def websocket_rpc_endpoint(websocket: WebSocket):
    async with endpoint.broadcaster:
        await endpoint.main_loop(websocket)


async def events(topic):
    await endpoint.publish([topic], {'msg': 'fastapi'})


@app.get('/trigger')
async def trigger_events():
    asyncio.create_task(events('event'))


app.include_router(router)

if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)

```

### client.py
```python
import asyncio
from fastapi_websocket_pubsub import PubSubClient


async def on_events(data, topic):
    print(f"running callback for {data} {topic}!")


async def main():
    client = PubSubClient(["event"], callback=on_events)
    client.start_client(f"ws://localhost:8000/pubsub")
    await client.wait_until_done()


if __name__ == '__main__':
    asyncio.run(main())
```

### test.py
```python
import asyncio
import random
from fastapi_websocket_pubsub import PubSubEndpoint

endpoint = PubSubEndpoint(broadcaster='redis://localhost:6379')


async def run():
    while True:
        await endpoint.publish('event', {'msg': random.randint(1, 100)})
        await asyncio.sleep(1)


if __name__ == '__main__':
    asyncio.run(run())
```

### js
```js
let ws = new WebSocket("ws://127.0.0.1:8000/pubsub")

let payload_str = JSON.stringify({request: {method: "subscribe", "arguments": {"topics": ["event"]}}})

ws.onmessage = (event) => {
    console.log('get event data', JSON.parse(event.data))
}

ws.send(payload_str)
```

**Reply by orweis** [reply]
Hi @hgalytoby - as I explained earlier you can't expect this to work without implementing the **entire** set of protocols, both RPC and PUB/SUB.

You need to write a full version of [RPC](https://github.com/permitio/fastapi_websocket_rpc), and a full version of [pub_sub ](https://github.com/permitio/fastapi_websocket_pubsub)  in JS, otherwise things would continue to not work correctly. 

In this case for example you're ignoring the fact that the subscribe function on the pub/sub server expects the RPC client on the other side to implement the [notify method](https://github.com/permitio/fastapi_websocket_pubsub/blob/49b8dfc7eac055d264ea492d559cee975d5574a6/fastapi_websocket_pubsub/rpc_event_methods.py#L86). Without the notify method sending an RPC response back, which the [RPC server is waiting for](https://github.com/permitio/fastapi_websocket_rpc/blob/0edd769c58dc3df47f222b5133ac5606ef53c2f7/fastapi_websocket_rpc/rpc_channel.py#L263) the server would hang until a configurable timeout (which is forever by default).


!! NOTE !! :
I really do want to help you, but you can't expect to do this with out being serious about it, and implementing a full JS version of the protocols.
I am sorry - but to be respectful of everyone's time - I won't respond to further questions unless you fully try to implement the RPC and pub/sub protocols.

**Reply by hgalytoby** [reply]
Thanks for taking the time to read, I get it.
